### ISSUE TITLE
CA2254 Should not apply when log message template is a resource string.

### ISSUE THREAD

**Issue by ghost** [dotnet/roslyn-analyzers#5626]
**Diagnostic ID**: [CA2254]

The code analyzer info message "The logging message template should not vary between calls to 'LoggerExtensions.LogDebug(ILogger, string?, params object?[])' appears when using a message template that is defined in a resource file.

Examples:
`logger.LogDebug(MyResources.MyMessageTemplate, param1, param2,...);`
`logger.LogDebug(localizer["MyMessageTemplate"], param1, param2,...);`

Resource strings are culture specific "constants", so it might make sense to treat properties of classes generated from .resx files and values of type LocalizedString as constants.

**Comment by nielstanghe** [comment]
Having the same issue since updating to Visual Studio 2022

**Comment by tbalakpm** [comment]
For me same issue since updated to .net 6.0

**Comment by jmihalich** [comment]
Same issue here, but i get it with a direct inline string with .net 6.0:

logger.LogDebug($"my message with {myparam}");

**Comment by Sour-Codes** [comment]
![image](https://user-images.githubusercontent.com/62261400/144894319-2f9c51d9-2006-4f46-a2eb-d6df4eae1b88.png)
I am targeting net7.0 and net6.0.  Builds just fine minus the warning.

**Comment by systemidx** [comment]
I also get this on a basic global exception handler middleware for MVC.

`public class GlobalExceptionMiddleware
    {
        private readonly RequestDelegate _next;
        private readonly ILogger<GlobalExceptionMiddleware> _logger;

        public GlobalExceptionMiddleware(RequestDelegate next, ILogger<GlobalExceptionMiddleware> logger)
        {
            _next = next;
            _logger = logger;
        }

        public async Task Invoke(HttpContext httpContext)
        {
            try
            {
                await _next(httpContext);
            }
            catch (Exception ex)
            {
                _logger.LogError(ex, ex.Message);
                throw;
            }
        }
    }`

**Comment by arthuredgarov** [comment]
@systemidx, the `ex.Message` is not a static string, so the warning you are getting is not a false positive.

**Comment by systemidx** [comment]
@arthuredgarov You're right, it's not. It's not a 100% analogue to the OP's issue of providing a static string from a resource file, but I still get CA2254 for some reason when pushing `ex` and `ex.Message` to ILogger.LogError. This didn't happen before the last VS2022 update on my end.

**Comment by arthuredgarov** [comment]
@systemidx, well, yes, that's what this analyzer rule stands for. You are providing a non-static string as the `message` parameter and getting the corresponding warning. The `message` parameter is intended to be a format string, you can check that in the method summary. By the way, if you are using the default implementation of the ILogger interface - you should not pass the `ex.Message` since the exception message will be logged and displayed from the `ex` instance you pass as the first parameter. Doing the way you do will lead to exception message duplicating in your logs. Instead, you could provide a more informative message describing the exception that occurred. If you don't want to do that, you could just define your own extension method that will call `_logger.LogError(ex, "");`. for example:

```
public static void LogException(this ILogger logger, Exception ex)
    => logger.LogError(ex, "");
```

**Comment by arthuredgarov** [comment]
@Sour-Codes, the same for you. Please check my comment above.

**Comment by simkin2004** [comment]
@arthuredgarov : In a strict sense of the definition "static", you are correct.  

Pragmatically, the point of this analyzer is to determine whether the code is building a string to use in the logger.  A string defined in a Resource file is a static string in the sense that its value -for a given Culture- cannot change at run-time.  

Providing log entries for the developer/ops person in the appropriate language is important, unless you believe that Microsoft should deliver all of its log messages in English only or Chinese only or French only.

Until this analyzer message provides the pragmatic value of preventing developers from building strings to pass into loggers, it will be suppressed as a false positive by many in the community.

**Comment by SoftCircuits** [comment]
I get the same warning with the following.

```cs
Logger.LogInformation($"**** Starting {ApplicationInfo.Description} ****");
```

What's strange is that I get the exact same warning this.

```cs
string start = $"**** Starting {ApplicationInfo.Description} ****";
Logger.LogInformation(start);
```

But I don't get the warning with this.

```cs
Logger.LogInformation("**** Starting ApplicationInfo.Description ****");
```

The error message makes no sense to me. I can't make heads or tails of this.

**Comment by kcaswick** [comment]
> I get the same warning with the following.
> 
> ```cs
> Logger.LogInformation($"**** Starting {ApplicationInfo.Description} ****");
> ```

The warning wants you to change to this:
```cs
Logger.LogInformation("**** Starting {0} ****", ApplicationInfo.Description);
```
That's harder to read though.. my preference would be the compiler turns string interpolation into a format string and parameters when passed to a function like this, or with some extra character.

**Comment by SoftCircuits** [comment]
@kcaswick Thanks for an explanation of this warning. The wording doesn't seem to say what you said. They should have you write the warning. ;-)

Yeah, in this case the parameter has absolutely nothing to do with anything I want to be part of logging. I'm just trying to use string interpellation to build a string. The compiler is being too smart here.

**Comment by Trolldemorted** [comment]
>The warning wants you to change to this:

Any idea why the latter would be better? It is not as readable as the former and they should be equally fast.

**Comment by kcaswick** [comment]
> Any idea why the latter would be better? It is not as readable as the former and they should be equally fast.

Probably this: (longer explanation at https://stackoverflow.com/a/49434820/20068)
> If you use the {0} format, the argument won't get translated into a string until after the logging framework has decided whether or not to emit this message, whereas the {myVar} format is translated to a string before the function is entered.

Also it is a problem if you use structured rather than plain text logging: https://habr.com/en/post/591171/

**Comment by Trolldemorted** [comment]
>Probably this: (longer explanation at https://stackoverflow.com/a/49434820/20068)

Thanks! Could we get the gist of that into the warning? I think I am not the only one who reacted with "but why though" at first.

**Comment by maryamariyan** [comment]
As with `LogError`, `LogDebug`, `LogInformation`, etc. the best practice is to move away from using interpolated strings as input string. So precisely the way explained in https://github.com/dotnet/roslyn-analyzers/issues/5626#issuecomment-1018003922, 

```c#
Logger.LogInformation($"**** Starting {ApplicationInfo.Description} ****");
```

needs to change to 

```c#
Logger.LogInformation("**** Starting {0} ****", ApplicationInfo.Description);
```

The purpose of `CA2254` is diagnostics is to not have mutable strings logged. So an input like:

```c#
logger.LogDebug($"my message with {myparam}");
```

needs to change to 
```c#
logger.LogDebug("my message with {MyParamInMessageTemplate}", myparam);
```


and as with the original issue:

```c#
logger.LogDebug(MyResources.MyMessageTemplate, param1, param2,...);
logger.LogDebug(localizer["MyMessageTemplate"], param1, param2,...);
```

**Comment by SoftCircuits** [comment]
@maryamariyan I understand the intent. But not every string I log needs splitting up into different fields. Sometimes I'm just trying to log a string from different sources. There is absolutely no benefit that I can see from treating the application description any differently than the rest of the string.

**Comment by maryamariyan** [comment]
> @maryamariyan I understand the intent. But not every string I log needs splitting up into different fields. Sometimes I'm just trying to log a string from different sources. There is absolutely no benefit that I can see from treating the application description any differently than the rest of the string.

Thanks for the feedback. If it feels onerous to do changes on the developer side in some cases then maybe the diagnostics needs to be revisited. e.g. perhaps we could consider to change the diagnostic severity from warning to information. Your case or the others represented above are not really false positives. The analyzer was supposed to catch them intentionally.

**Comment by naughtyGitCat** [comment]
that feels make c# logging code verbose like java, I think it makes code hard to read, willing to rollback

**Comment by arthuredgarov** [comment]
@maryamariyan,

> or the others represented above are not really false positives.

OP report as well? I mean, in case I have the following code I will get CA2254 too:

```
//Somewhere in .resx file
MyResource.Greeting = "Welcome to {city} {province}!"

//Logging itself
LogInformation(MyResource.Greeting, city, province)
```

**Comment by maryamariyan** [comment]
The reason we don't recommend using interpolated strings with logging APIs is the message templates are meant to help with producing structured log messages. This doc https://messagetemplates.org/ goes over message templates.

---

> that feels make c# logging code verbose like java, I think it makes code hard to read, willing to rollback

We're open to revisiting this and would like to better understand developer expectations before proceeding with next actions here. To do this, let's go through another logging example (this time not using LogError/LogWarning but the new logging source generator feature instead). Consider the new logging source generator feature presented in [here](https://docs.microsoft.com/en-us/dotnet/core/extensions/logger-message-generator), with the sample usage code below: 

```c#
public static partial class Log
{
    [LoggerMessage(
        EventId = 0,
        Level = LogLevel.Critical,
        Message = "Could not open socket to `{hostName}`")]
    public static partial void CouldNotOpenSocket(
        ILogger logger, string hostName);
}
```

With this sample usage, you'd see the `Message` is the message template and arguments to the method would be the variables you pass to it. The message template that the developer places in the `LoggerMessage` annotation as `Message` is what is supposed to **not vary between calls**. This way the message template gets written once and for every usage the developer only calls this method passing down any variables as argument to the log method.

---

@Youssef1313 @333fred what is your take on how the analyzer should behave for the use case below presented in comment https://github.com/dotnet/roslyn-analyzers/issues/5626#issuecomment-1034391219 above?

```c#
//Somewhere in .resx file
MyResource.Greeting = "Welcome to {city} {province}!"

//Logging itself
LogInformation(MyResource.Greeting, city, province)
```

**Comment by MisinformedDNA** [comment]
The analyzer rule is useful IMO, but there should be a code fix and improved [documentation](https://docs.microsoft.com/en-us/dotnet/fundamentals/code-analysis/quality-rules/ca2254). This is way too little information:

![image](https://user-images.githubusercontent.com/1784452/155620672-d61c2ee5-22e6-4512-88bd-11f0849dc99c.png)

**Comment by mavasani** [comment]
@gewarren for documentation improvement

**Comment by StevenQuickGS1** [comment]
It warns for string templates that are defined as readonly static which doesn't seem useful 

```
static readonly string myTemplate = "{Something}: {SomethingElse}, {@Data}";
logger.LogInformation(myTemplate, "Flux Capacitor", "Modulate", new { Power = 35, Dimension = "Next"});
```

**Comment by simkin2004** [comment]
Multiple developers have provided reasons why the functionality that changed from one version of Visual Studio to another has flaws that cause this rule to be less usable.  

I would argue that the Principle of Least Astonishment should be followed with future BREAKING changes like this.

If you want to make a change like this, create a new rule and let developers decide which one makes the most sense for their codebase.

**Comment by mike1880** [comment]
It seems to me that trying to force everyone down the structured logging path, whether or not they need it or want it, is the wrong decision. By all means document how structured logging can be facilitated by doing things a particular way, but don't ram it down our throats if we don't want it. I don't need it and I dislike having to set up a global suppression file just because I don't happen to need a particular feature.

**Comment by ScottyMac52** [comment]
> Same issue here, but i get it with a direct inline string with .net 6.0:
> 
> logger.LogDebug($"my message with {myparam}");
> 
> ![image](https://user-images.githubusercontent.com/15250728/143793501-d812dccb-0166-4536-b414-87d6e6a824cc.png)

Technically what you are doing is the same. The Warning will go away if the message is a string literal. Somebody made a VERY wrong decision to enforce something like this.

**Comment by battarro** [comment]
Structured logging is a much better approach to logging than just writing a string out. 

The workaround of simply ignoring or downgrading the severity of the issue on your visual studio seems like a good compromise that can accomplish both goals.

I disagree that it was a bad decision to warn us about this, warning us is fine, then we can do what the warning as we wish.

**Comment by qsdfplkj** [comment]
You should use structured message logging and not string interpolation:

    logger.Log($"The value: {value}");

Should be 

    logger.Log("The value: {anyname}", value);

But if you insist on using string interpolation put everything in the arguments.

    logger.Log("{0}",  $"The value: {value}")

Note that the template should be a *constant*  not a readonly static 

You could also use the more performant LoggerMessage or LoggerMessageAttribute approaches.

**Comment by bronumski** [comment]
As someone who bangs the drum about not using `string.format` and string interpolation in log messages because you are preventing the framework from doing its job I can now start to relent on the use of string interpolation. With the use of the `InterpolatedStringHandlerAttribute` this is now possible and we can not only defer the rendering of the string and check if the log level is enabled but we can also populated the structured logs with the interpolated string arguments. This is not something I have come up with but is made clear in the Microsoft [doc articles](https://docs.microsoft.com/en-us/dotnet/csharp/whats-new/tutorials/interpolated-string-handler) . The reason I am here is because I am getting the same warning in my implementation of extension methods around the `ILogger` which allows you to do this:

```
logger.LogInformation($"We can now do this {someCode}");

logger.LogInformation($"And this {1 + 1:Calculated}");

logger.LogInformation($"And also this {() => ExpensiveMethodCall(TimeSpan.FromSeconds(1)):Expensive}");

logger.LogInformation($"And not worry about {UnsafeMethodCall:Unsafe}");
```

I am getting this warning in the implementation of this extension method and given the implementation I don't think it is valid:

```
private static void Log(this ILogger logger, (string template, object[] arguments)? interpolatedValues, LogLevel logLevel, Exception? ex)
    {
        if (!interpolatedValues.HasValue) return;
        var (template, arguments) = interpolatedValues.Value;
        logger.Log(logLevel, ex, template, arguments);
    }
```

https://gist.github.com/bronumski/8ffbd37b9446794be4acc5a9a810c776

**Comment by DustinCampbell** [comment]
For a personal project, I ended up creating a more efficient set of "Log*" extension methods that use an interpolated string handler. This should be much more efficient and avoid allocating a `StringBuilder` for each log method. Sharing here for anyone interested.

```C#
using System.Runtime.CompilerServices;
using System.Text;

namespace Microsoft.Extensions.Logging;

#pragma warning disable CA2254 // Template should be a static expression

public static class LoggerInterpolatedStringExtensions
{
    private const int MaxCapacity = 512;

    [ThreadStatic]
    private static readonly StringBuilder t_builder = new();

    [InterpolatedStringHandler]
    public ref struct LogInterpolatedStringHandler
    {
        private StringBuilder.AppendInterpolatedStringHandler _handler;

        public LogInterpolatedStringHandler(int literalLength, int formattedCount)
        {
            _handler = new StringBuilder.AppendInterpolatedStringHandler(literalLength, formattedCount, t_builder);
        }

        public string ToStringAndClear()
        {
            var builder = t_builder;
            string result = builder.ToString();
            builder.Clear();
            
            if (builder.Capacity > MaxCapacity)
            {
                builder.Capacity = MaxCapacity;
            }

            _handler = default;
            return result;
        }

        public void AppendLiteral(string value)
            => _handler.AppendLiteral(value);

        public void AppendFormatted<T>(T value)
            => _handler.AppendFormatted(value);

        public void AppendFormatted<T>(T value, string? format)
            => _handler.AppendFormatted(value, format);

        public void AppendFormatted<T>(T value, int alignment)
            => _handler.AppendFormatted(value, alignment, format: null);

        public void AppendFormatted<T>(T value, int alignment, string? format)
            => _handler.AppendFormatted(value, alignment, format);

        public void AppendFormatted(ReadOnlySpan<char> value)
            => _handler.AppendFormatted(value);

        public void AppendFormatted(ReadOnlySpan<char> value, int alignment = 0, string? format = null)
            => _handler.AppendFormatted(value, alignment, format);

        public void AppendFormatted(string? value)
            => _handler.AppendFormatted(value);

        public void AppendFormatted(string? value, int alignment = 0, string? format = null)
            => _handler.AppendFormatted(value, alignment, format);

        public void AppendFormatted(object? value, int alignment = 0, string? format = null)
            => _handler.AppendFormatted(value, alignment, format);
    }

    public static void LogCritical(this ILogger logger, ref LogInterpolatedStringHandler message)
        => logger.Log(LogLevel.Critical, ref message);

    public static void LogCritical(this ILogger logger, EventId eventId, ref LogInterpolatedStringHandler message)
        => logger.Log(LogLevel.Critical, eventId, ref message);

    public static void LogCritical(this ILogger logger, Exception? exception, ref LogInterpolatedStringHandler message)
        => logger.Log(LogLevel.Critical, exception, ref message);

    public static void LogCritical(this ILogger logger, EventId eventId, Exception? exception, ref LogInterpolatedStringHandler message)
        => logger.Log(LogLevel.Critical, eventId, exception, ref message);

    public static void LogError(this ILogger logger, ref LogInterpolatedStringHandler message)
        => logger.Log(LogLevel.Error, ref message);

    public static void LogError(this ILogger logger, EventId eventId, ref LogInterpolatedStringHandler message)
        => logger.Log(LogLevel.Error, eventId, ref message);

    public static void LogError(this ILogger logger, Exception? exception, ref LogInterpolatedStringHandler message)
        => logger.Log(LogLevel.Error, exception, ref message);

    public static void LogError(this ILogger logger, EventId eventId, Exception? exception, ref LogInterpolatedStringHandler message)
        => logger.Log(LogLevel.Error, eventId, exception, ref message);

    public static void LogWarning(this ILogger logger, ref LogInterpolatedStringHandler message)
        => logger.Log(LogLevel.Warning, ref message);

    public static void LogWarning(this ILogger logger, EventId eventId, ref LogInterpolatedStringHandler message)
        => logger.Log(LogLevel.Warning, eventId, ref message);

    public static void LogWarning(this ILogger logger, Exception? exception, ref LogInterpolatedStringHandler message)
        => logger.Log(LogLevel.Warning, exception, ref message);

    public static void LogWarning(this ILogger logger, EventId eventId, Exception? exception, ref LogInterpolatedStringHandler message)
        => logger.Log(LogLevel.Warning, eventId, exception, ref message);

    public static void LogInformation(this ILogger logger, ref LogInterpolatedStringHandler message)
        => logger.Log(LogLevel.Information, ref message);

    public static void LogInformation(this ILogger logger, EventId eventId, ref LogInterpolatedStringHandler message)
        => logger.Log(LogLevel.Information, eventId, ref message);

    public static void LogInformation(this ILogger logger, Exception? exception, ref LogInterpolatedStringHandler message)
        => logger.Log(LogLevel.Information, exception, ref message);

    public static void LogInformation(this ILogger logger, EventId eventId, Exception? exception, ref LogInterpolatedStringHandler message)
        => logger.Log(LogLevel.Information, eventId, exception, ref message);

    public static void LogDebug(this ILogger logger, ref LogInterpolatedStringHandler message)
        => logger.Log(LogLevel.Debug, ref message);

    public static void LogDebug(this ILogger logger, EventId eventId, ref LogInterpolatedStringHandler message)
        => logger.Log(LogLevel.Debug, eventId, ref message);

    public static void LogDebug(this ILogger logger, Exception? exception, ref LogInterpolatedStringHandler message)
        => logger.Log(LogLevel.Debug, exception, ref message);

    public static void LogDebug(this ILogger logger, EventId eventId, Exception? exception, ref LogInterpolatedStringHandler message)
        => logger.Log(LogLevel.Debug, eventId, exception, ref message);

    public static void LogTrace(this ILogger logger, ref LogInterpolatedStringHandler message)
        => logger.Log(LogLevel.Trace, ref message);

    public static void LogTrace(this ILogger logger, EventId eventId, ref LogInterpolatedStringHandler message)
        => logger.Log(LogLevel.Trace, eventId, ref message);

    public static void LogTrace(this ILogger logger, Exception? exception, ref LogInterpolatedStringHandler message)
        => logger.Log(LogLevel.Trace, exception, ref message);

    public static void LogTrace(this ILogger logger, EventId eventId, Exception? exception, ref LogInterpolatedStringHandler message)
        => logger.Log(LogLevel.Trace, eventId, exception, ref message);

    public static void Log(this ILogger logger, LogLevel logLevel, ref LogInterpolatedStringHandler message)
    {
        if (!logger.IsEnabled(logLevel))
        {
            return;
        }

        logger.Log(logLevel, message.ToStringAndClear(), Array.Empty<object?>());
    }

    public static void Log(this ILogger logger, LogLevel logLevel, EventId eventId, ref LogInterpolatedStringHandler message)
    {
        if (!logger.IsEnabled(logLevel))
        {
            return;
        }

        logger.Log(logLevel, eventId, message.ToStringAndClear(), Array.Empty<object?>());
    }

    public static void Log(this ILogger logger, LogLevel logLevel, Exception? exception, ref LogInterpolatedStringHandler message)
    {
        if (!logger.IsEnabled(logLevel))
        {
            return;
        }

        logger.Log(logLevel, exception, message.ToStringAndClear(), Array.Empty<object?>());
    }

    public static void Log(this ILogger logger, LogLevel logLevel, EventId eventId, Exception? exception, ref LogInterpolatedStringHandler message)
    {
        if (!logger.IsEnabled(logLevel))
        {
            return;
        }

        logger.Log(logLevel, eventId, exception, message.ToStringAndClear(), Array.Empty<object?>());
    }
}
```

**Comment by crozone** [comment]
> As someone who bangs the drum about not using `string.format` and string interpolation in log messages because you are preventing the framework from doing its job I can now start to relent on the use of string interpolation. With the use of the `InterpolatedStringHandlerAttribute` this is now possible and we can not only defer the rendering of the string and check if the log level is enabled but we can also populated the structured logs with the interpolated string arguments. This is not something I have come up with but is made clear in the Microsoft [doc articles](https://docs.microsoft.com/en-us/dotnet/csharp/whats-new/tutorials/interpolated-string-handler) .

I'm very surprised the `ILogger` methods don't do this by default. Sure, interpolated strings aren't as good for structured logging, but I would have assumed that this would have been done as an optimization for late/lazy evaluation of the interpolated string.

**Comment by RickyLin** [comment]
I'm okay with `_logger.LogError(ex, "");` , it makes sense and does remove the CA2254.
But I'm curious why the following code don't remove the CA2554?

```
_logger.LogError(ex, null);
_logger.LogError(ex, string.Empty);
```

**Comment by qsdfplkj** [comment]
> I'm okay with `_logger.LogError(ex, "");` , it makes sense and does remove the CA2254. But I'm curious why the following code don't remove the CA2554?
> 
> ```
> _logger.LogError(ex, null);
> _logger.LogError(ex, string.Empty);
> ```

Null and String.empty is are not a constant. Where you might thing the null could be ambiguous but what if it is a variable that is null.

**Comment by MisinformedDNA** [comment]
`null` is a constant.

```csharp
const string Null = null;
const string Empty = "";
const string StringEmpty = string.Empty;  // error here only
```

**Comment by skrabbe001** [comment]
Now please stop wasting most of the developers time, and make you framework usable. Why must we again and again waste our time on issues, that shouldn't even be an issue.

I have never in my life had to check github, to find out what a function actually does, but I guess that is over. Most of the manuals don't even describe what the function in question does.

C# supports interpolated strings. Either remove it from the language. Or make logging functions that accept them. Christ.

**Comment by bronumski** [comment]
@skrabbe001 Easy tiger! I sense your frustration but a rant is not going to help anyone. Software is a collaborative and learning processes and not everything is going to be the way you want it. A seemingly innocuous change to you might have bigger ramifications elsewhere in a publicly used library and cannot be changed on a whim. The fact that dotnet is open source and we as a community are able to have open conversations to influence the direction the tools we use take is awesome and a far cry from the early years of .net .

**Comment by qsdfplkj** [comment]
> `null` is a constant.
> 
> ```cs
> const string Null = null;
> const string Empty = "";
> const string StringEmpty = string.Empty;  // error here only
> ```

Const string null is still null. And the logger accepts a nullable string. But then it Triggers ca2254? But just using null directly causes the ca2254 logger.error(ex, null).

**Comment by mpaperno** [comment]
LOL... So in .NET 5 I used interpolated strings in logging (and elsewhere), 'cuz hey, it's easier to read and write.,,, cool!  Now turns out that's not cool?  Obviously I missed the memo.  

And also turns out it doesn't like numbered format placeholders either... no, they need to be named.... otherwise you get a CA2253 instead.

BTW I don't care about "structured logging" (in most cases), just need to get some info into a file or console.  

As others have suggested, if "structured logging" is so important that it warrants a warning by default, why not have the compiler just do the "right thing"? How hard is it to turn `($"My {message}")` into `("My {message}", message)`?

Kinda funny, C# can't seem to decide if it's an easy-to-use "we've got your back" language or some rigid framework.  Oh well, off to add yet another warning to my ignore list.

BTW I certainly don't mean to disparage anyone at all here,.. this kind of stuff is thankless work and I'm sure everyone is doing their best and with best intentions. I honestly find it more amusing than anything.

Cheers,
-Max

**Comment by mickelsonmichael** [comment]
> As others have suggested, if "structured logging" is so important that it warrants a warning by default, why not have the compiler just do the "right thing"? How hard is it to turn `($"My {message}")` into `("My {message}", message)`?

I'm no expert in this domain, but I believe the flaw with this would be inconsistent naming. Your variable names now need to conform to a consistent and standard style in order to be interpreted properly by some log analyzer. The log analyzer won't be able to nicely get all the "message" data out if at some point a variable was named `msg` or `logMessage` instead, and you would need to know about and account for all those instances where it changed. The constant strings promote that consistent naming by not being dependent on the variable names in the scope.

And even though you may not use the capabilities of structured logging right now, there's no telling what you may be using in the future, and it would be beneficial to already have the ability to switch over without changing any code. In my opinion, this would be the more pragmatic approach.

I think for any legitimate cases where you truly don't need structured logging, you can always just [disable the particular rule](https://docs.microsoft.com/en-us/dotnet/fundamentals/code-analysis/suppress-warnings).

**Comment by qsdfplkj** [comment]
You are partially right about the template changing with different identifiers but when you have the same statement  but have different values you can find those based on the templates (from the interpolated string).

The “compiler” approach is similar to what I suggested before. You will only have one place where you can suppress the warning because you know you are using a templates from the interpolated string.

If you have multiple place in your code you will probably not be using similar interpolated strings. Because if you do you would have to apply some common practice and store the template in a variable (hence the constant) maybe you can use some extensions methods similar  as the loggermessage.define approach.

**Comment by mpaperno** [comment]
Apologies for (maybe) veering more off-topic...  If one needs highly structured data for later analysis, why not just log it into a database? Is that what "structured logging" is trying to recreate?

Having said that, I did discover/realize, thanks to this issue, that the Serilog Console logger will color-code replacements based on their data type... so that's pretty neat (even in .NET 5).

But the actual reason I'm switching my code to non-interpolated log strings, which I should have realized a lot sooner (I thought .net "had my back" lol), is contained in this snip from the previously-quoted [SO answer](https://stackoverflow.com/questions/37985012/which-of-one-from-string-interpolation-and-string-format-is-better-in-performanc/49434820#49434820):

> If you use the {0} format, the argument won't get translated into a string until after the logging framework has decided whether or not to emit this message, whereas the {myVar} format is translated to a string before the function is entered. 

I certainly prefer "lazy" logging any day.

**Comment by qsdfplkj** [comment]
In that case checkout loggermessage.define and the loggermessageattribute.

**Comment by battarro** [comment]
> Apologies for (maybe) veering more off-topic... If one needs highly structured data for later analysis, why not just log it into a database? Is that what "structured logging" is trying to recreate?
> 
> Having said that, I did discover/realize, thanks to this issue, that the Serilog Console logger will color-code replacements based on their data type... so that's pretty neat (even in .NET 5).
> 
> But the actual reason I'm switching my code to non-interpolated log strings, which I should have realized a lot sooner (I thought .net "had my back" lol), is contained in this snip from the previously-quoted [SO answer](https://stackoverflow.com/questions/37985012/which-of-one-from-string-interpolation-and-string-format-is-better-in-performanc/49434820#49434820):
> 
> > If you use the {0} format, the argument won't get translated into a string until after the logging framework has decided whether or not to emit this message, whereas the {myVar} format is translated to a string before the function is entered.
> 
> I certainly prefer "lazy" logging any day. IMHO any warnings should point _that_ out.

Database logging is not a solution. It will degrade your application performance to a crawl and it will increase your actual hosting costs and network traffic. 

Imagine a simple scenario where your systems picks up a record to do from database, process it and saves the output in another table. That is only two database tripa..trips... now if your process is doing anything other than 2 lines of codes, you will start logging parameters and possible values and temporary states. If you have to go to the database every time to save that log, just forget it, you are done. 
On most of my apps, I do exponentially more log calls than database calls, because on cloud computing you can't debug on the server, so you have to log methods and their parameters, so you can then reproduce any errors.
Aws and azure offer high throughput logging mechanism that won't hurt your performance.

**Comment by battarro** [comment]
Back to the topic. Don't change the current behavior. Structured logging is the better solution and the compiler should hint at people so they change their habits little by little.

**Comment by qsdfplkj** [comment]
One thing I personally I like to do is:

    logger.LogError(ex, ex.Message);

This of course  triggers the warning. Rewriting this to avoid the warning kind of defeats the purpose of the warning.

   logger.LogError(ex, "ExceptionMessage: {exceptionMessagse}", ex.Message);

**Comment by arthuredgarov** [comment]
> One thing I personally I like to do is:
> 
> ```
> logger.LogError(ex, ex.Message);
> ```

@qsdfplkj, please look at my [comment](https://github.com/dotnet/roslyn-analyzers/issues/5626#issuecomment-993621070) above.

**Comment by bronumski** [comment]
@qsdfplkj as @arthuredgarov said, let the framework do the job for you. It knows what to do with an exception, you can set up log templates to include the message in the log. Better yet structured logging libraries (serilog) and tools (seq) can render the exception much more clearly. Additionally if you are logging an exception at the debug level and have disabled debug the app will not waste time getting the log message in order to render it.

**Comment by qsdfplkj** [comment]
It was me bring clever/lazy to write a proper log message. But it makes sense now.

**Comment by skrabbe001** [comment]
Using an interpolated string IS structured.

I repeat: We are here to develop software, not to work around quirks, that we can only find out about, by checking the bug reports to the manual page.

Calling bad manual pages (where you have to check the bug reports to the manual page to find out what the function does or error message means), writing log systems that work against the structured messages of the language (ie. Interpolated string) progress is …. Far fetched.

**Comment by skrabbe001** [comment]
>but a rant is not going to help anyone

An interpolated string IS structured.
So structured messages are built into the language.
That developers should not have to check the bug reports for a manual page, to find out what the manual page means.

Frameworks/libraries and manual pages are created to help the developer spend _less_ time programming …. Or am I mistaken?

If you consider that a rant, good for you. Do that, instead of addressing the issues.

**Comment by skrabbe001** [comment]
Also, I fail to see how your message is helpful.

> far cry from the early years of .net .

And still a far cry away from the consistency of UNIX and it’s manual pages from more than 30 years ago.

**Comment by eerhardt** [comment]
Is anyone seeing CA2254 as an actual **_warning_**? The default severity for that rule is "info", which means it shows up under "Messages" in your error list, but it shouldn't be emitted as an actual warning:

![image](https://user-images.githubusercontent.com/8291187/183110657-720f96bf-dabc-4446-8d18-3186304c38c7.png)

If you are getting warnings, can you post a picture of it, and what your editor config (if any) looks like?

**Comment by qsdfplkj** [comment]
So why now #pragma warning disable CA2254 causes an IDE0097 remove unnecessary suppression?

**Comment by battarro** [comment]
pragma warning only deals with warnings.

To be clear, you are still doing something considered inappropriate, that is why it is informing you.

**Comment by qsdfplkj** [comment]
I don't understand. This pragma is just added when suppressing it in code and now suddenly it is unnecessary. And I've no clue what changed.

**Comment by ygoe** [comment]
I'm not sure what this message does at all, because it even applies when there are no parameters at all. There is nothing to resolve here, I cannot do anything to remove that message:

    logger.LogError(ex, ex.Message);

**Comment by changhuixu** [comment]
You can do this `_logger.LogError(ex, "{m}", ex.Message);`
Though, I agree that it doesn't look as elegant as `_logger.LogError(ex, ex.Message);`

> I'm not sure what this message does at all, because it even applies when there are no parameters at all. There is nothing to resolve here, I cannot do anything to remove that message:
> 
> ```
> logger.LogError(ex, ex.Message);
> ```
> 
> PS: This is a simplified repro example. My actual code uses an extension method that returns a more useful message:
> 
> ```
> logger.LogError(ex, ex.GetRecursiveMessage());
> ```

**Comment by ygoe** [comment]
> You can do this `_logger.LogError(ex, "{m}", ex.Message);` Though, I agree that it doesn't look as elegant as `_logger.LogError(ex, ex.Message);`

Sure I could, but why would I want to do that? Isn't the simplest solution the best? What's the benefit of this additional complexity at all?

**Comment by mickelsonmichael** [comment]
@ygoe in your contrived example, the `ex.Message` will be logged anyway by most common formats, meaning you will be outputting the message twice.

I'm not sure what your `ex.GetRecursiveMessage()` does, but it would make a lot of sense to make an `ILogger.LogRecursiveException(Exception)` extension method to clean that code up. Then you could either avoid the flag entirely (depending on the implementation), or you can simply ignore it on the singular line using `#pragma warning disable CA2254 // Template should be a static expression`, and won't need to repeat yourself.

**Comment by ygoe** [comment]
@mickelsonmichael That method collects the messages of inner exceptions all into a single string. Often, the message of the caught exception tells nothing more than that there was any error and details can be found in the inner exceptions. Such a log message is then useless. To avoid that, my extension method generates more useful messages.

I also have other examples, like:

    logger.LogError(ex, "The XYZ operation could not be completed. " + ex.GetRecursiveMessage());

This would basically add another level of error messages on top of the exception stack. It gives them a context that they might not have, and makes it really easy to understand an error from a single log entry. (I also log the exception details but sometimes they are simply unavailable from the log, depending on how it's accessed.)

I guess I'll shut down that warning on a global level. At least until somebody showed me a good example why I would want to follow it. It still seems pointless and arbitrary to me and provides no benefit at all. It's even less readable and more error-prone than an interpolated string which puts the parameters right where they belong, instead of counting arguments and eventually mixing them up.

BTW, setting the severity to NONE from the quick actions has no effect at all. No files were changed and the warning doesn't go away. Is there a problem with the code fixes? Typing the suggested fix manually into my .editorconfig file resolves the issue immediately:

    # Don't suggest using ILogger placeholders and arguments
    dotnet_diagnostic.CA2254.severity = none

I'm going to include that in my global defaults.

**Comment by Dean-NC** [comment]
I know log methods expect a template, but I don't think it should generate a warning for a runtime string that isn't one. If you want to use a template, great, but it would be nice to log a message using a variable or method call without a warning. Example: `logger.LogInformation(GetMessageToLog());` but this gives the warning.

**Comment by simkin2004** [comment]
90% of the comments on this thread are tangential to the original request.

I (and the original poster) still have not received a response about how to go about handling logging in different languages using Resource files and still complying with this rule.  

If that is not possible, then update the documentation to make sure anyone with i18n or l10n requirements for that logging are properly informed to disable this rule.

**Comment by tarekgh** [comment]
@simkin2004 we are keeping this issue open to see if we can do something for the resource's cases.

**Comment by asimontsev** [comment]
I like the idea of structured logging, however, I believe it should be applied only to certain log levels, say, Information and higher. I don't think that it is worthy to do it for Debug and Trace level.

In my understanding, using structured logs are like creating a table with columns for each placeholder names. It means I should have consistent placeholder names through the entire project (which is absolutely fine), however it does not seem realistic for debug information (which may be quite unique for each particular method). As a result, I can end up with hundreds or thousands of "columns" in such "table".

Or my understanding is wrong here? If not, any chance to make it possible to fine tune the warning strategy depending on a log level?

**Comment by peabnuts123** [comment]
A lot has already been said in this thread so I'll just add what I feel hasn't been said: Logging is not always structured. For my Web API or console app, logs are just going into some kind of text output. I just want to output data so I can see what's happening. It seems aggressive to me to suggest all developers should use structured logging in all use-cases. It is my opinion that using a formatted string is less readable and therefore undesirable. When my project gets sophisticated enough that I'll need to actually query my logs, that trade-off will be more valuable, and I can make that change.

Since it's clear there is a lot of different opinions and needs represented by this issue, is there a way that I can disable this rule for my whole solution? It's obviously infeasible to disable it at every location I'm logging something. Some kind of entry in my csproj or something? I personally do not agree with it, so I should be able to opt-out.

**Comment by Youssef1313** [comment]
@peabnuts123 You can disable it for sure. In `.editorconfig`, add `dotnet_diagnostic.CA2254.severity = none`

**Comment by Gladskih** [comment]
Isn't the main goal of structured logging to have logs auto processible without parsing?
It is not about making an app to build a human readable string from data at all! If we need a data we shall just pass it as part of event. We shall not expect logging framework to include data back to message in event. Messing with templates was so wrong way that I cannot understand how the community has not rejected the idea yet (and for so long).

**Comment by dFarkhod** [comment]
While I understand that structured logging aims to improve performance and efficiency, it introduces a significant risk due to human error. Developers might forget to include one or more parameters, potentially leading to exceptions. In contrast, using string interpolation with $ ensures there’s no chance of such mistakes—it's straightforward and reliable.

I was thrilled when the $ operator was introduced, as it eliminated the cumbersome and error-prone String.Format(). Unfortunately, that joy was short-lived due to this frustrating warning about structured logging. Additionally, strings constructed with string interpolation are far easier to read.

Instead of imposing this inconvenient rule, the development environment should address the performance issues associated with logging. For example, it could avoid concatenating strings if logging is not enabled or if the log level does not meet the desired criteria. This responsibility shouldn’t fall on developers—let the system handle it seamlessly.

**Comment by ygoe** [comment]
It's probably not easy or even feasible to address these points with existing technology. First, the development environment (IDE?) can't do much about it. Maybe it could count and compare placeholders and passed parameters. Then, the log level is evaluated at a point where log call arguments have already been passed. So the only possible improvement here would be to accept a formattable string (I forgot the interface name again) so that the log method can generate the resulting string later and only when needed, as it does today with the placeholders.

But that would still lack the parameter names that exist with today's solution. Interpolated strings' placeholders are nameless. That might be something that could be addressed. Maybe with a proposed syntax like:

```cs
string userName = "Alice";
logger.LogInformation($"The user {Name: userName} logged in.");
```

This would assign the value of the string variable to the log message parameter "Name". Of course this collides with format specifiers. Using an `=` is also not possible because it would assign something in the code. Anyway, this is a whole new topic then.
